diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzy.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzy.h	2014-12-15 09:22:37.000000000 +0100
@@ -0,0 +1,82 @@
+ /*
+    OpalFuzzy.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 021101301, USA.
+ */
+ 
+ #ifndef OpalFuzzy_h_defined
+ #define OpalFuzzy_h_defined
+ 
+ @interface OpalFuzzyFunction : NSObject {
+ 	SEL pf;
+ 	long int x;
+ }
+-  (id) new:(SEL) pif; 
+- (void)perform:(id)c;
+-  (void) perform;
+-  (long int)x;
+-  (void)x:(long int)xx;
+ @end
+ 
+ @interface OpalFuzzyFunctionLittle : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionSlightly : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionVery : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionExtremely : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionVeryVery : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionMoreorLess : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionSomewhat : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionIndeed : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ 
+ @interface OpalFuzzy : NSObject {
+ 	
+ }
+ 
+ 
+ @end
+ 
+ #endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyDB.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyDB.h	2014-12-15 09:30:33.000000000 +0100
@@ -0,0 +1,40 @@
+ /*
+    OpalFuzzyDB.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #ifndef OpalFuzzyDB_h_defined
+ #define OpalFuzzyDB_h_defined
+ 
+ @interface OpalFuzzyDB : NSMutableDictionary {
+ }
+ - (id)new;
+ - (void)addToDict:(NSString*)key :(SEL)pif;
+ - (SEL)getWithKey:(NSString*)key;
+ - (void)reinitCapacity;
+ 
+ @end
+ 
+ #endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyManipulator.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyManipulator.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyManipulator.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyManipulator.h	2014-12-15 12:16:17.000000000 +0100
@@ -0,0 +1,61 @@
+/*
+   OpalFuzzyManipulator.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyManipulator_h_defined
+#define OpalFuzzyManipulator_h_defined
+
+@interface OpalFuzzyScreenManipulator : OpalFuzzyArgumentManipulator
+{
+}
+
+- (id) new:(id)screen;
+@end
+
+@interface OpalArgumentManipulator : NSObject 
+{
+	id _o;
+}
+- (id) new:(id)o;
+- (id) o;
+@end
+
+@interface OpalFuzzyPaintManipulator : OpalFuzzyManipulator
+{
+}
+
+- (id) new:(id)painter;
+@end
+
+@interface OpalFuzzyManipulator : NSObject
+{
+	id _o;
+}
+- (id) new:(id)o;
+- (id) o;
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyMatrix.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyMatrix.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyMatrix.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyMatrix.h	2014-12-15 08:41:41.000000000 +0100
@@ -0,0 +1,58 @@
+/*
+   OpalFuzzyMatrix.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyMatrix_h_defined
+#define OpalFuzzyMatrix_h_defined
+
+#import <Foundation/NSAffineTransform.h>
+
+#define MAKEMatrixElement(i,j) \
+	OpalFuzzyNode * m ## #i#j;
+
+/* FIXME extra handle or not */
+#define MAKEMatrix(w, h) \
+	typedef struct {
+		for (i = 0; i < h; i++) \
+			for (j = 0; j < w; j++) \
+				makeMatrixElement(i,j) \
+} _OpalMatrix;		
+
+@class _OpalMatrix;
+
+@interface OpalFuzzyMatrix : NSObject {
+	long int _w,_h;
+	_OpalMatrix *_matrix;	
+}
+
+- (id) new;
+- (id)setWidth:(long int)w height:(long int)h;
+- (BOOL)hasDimensionsBeenSet:(BOOL)b;
+- (void) calibrateDimensions;
+ 
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyNetwork.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNetwork.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyNetwork.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNetwork.h	2014-12-15 07:08:58.000000000 +0100
@@ -0,0 +1,37 @@
+/*
+   OpalFuzzyNetwork.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyNetwork_h_defined
+#define OpalFuzzyNetwork_h_defined
+
+@interface OpalFuzzyNetwork : NSObject {
+}
+- (id) new;
+ 
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyNode.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNode.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyNode.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNode.h	2014-12-15 10:51:27.000000000 +0100
@@ -0,0 +1,44 @@
+/*
+   OpalFuzzyNode.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyNode_h_defined
+#define OpalFuzzyNode_h_defined
+
+@class OpalFuzzyRuleSystem;
+
+@interface OpalFuzzyPaintNode : OpalFuzzyNode
+{}
+
+@end
+
+@interface OpalFuzzyNode : NSObject
+{
+	OpalFuzzyRuleSystem *_rules;
+}
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h	2014-12-15 12:21:48.000000000 +0100
@@ -0,0 +1,63 @@
+/*
+   OpalFuzzyPaintRuleSystem.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyPaintRuleSystem_h_defined
+#define OpalFuzzyPaintRuleSystem_h_defined
+
+@class OpalFuzzyRuleSystem;
+
+
+@interface OpalFuzzyPaintRuleSystem : OpalFuzzyRuleSystem 
+{
+	id _painter;
+}
+- (id) new:(id)painter;
+- (void)setPaintSystem:(id)paintsystem;
+- (void)painter;
+- (void)paintOn;
+
+- (void)paintOn:(id)painter;
+
+/*
+ * NOTE :
+ * Call the createManipulator method and you can paint on this
+ * painter system with rules from the super class.
+
+ * The PaintManipulator then hashes the method paintOn: in the dictionary
+ * and paints on the _painter of the dispatched class due to a call
+ * of e.g. parseFor:"paint extremely". 
+ *
+ * The super class parseFor:On: method can also be used as a rule (rules have
+ * a NSString key and a SEL in the form of performSelector:withObject:, so
+ * only a single argument.
+ */ 
+- (id) createManipulator;
+- (id) createArgumentManipulator;
+
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h	2014-12-15 12:18:53.000000000 +0100
@@ -0,0 +1,54 @@
+/*
+   OpalFuzzyRuleSystem.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyRuleSystem_h_defined
+#define OpalFuzzyRuleSystem_h_defined
+
+@class OpalFuzzyDB;
+
+@interface OpalFuzzyRuleSystem : NSObject
+{
+
+	OpalFuzzyDB *_rules;
+
+}
+- (id) new;
+- (void) addRule:(NSString*)r with:(SEL)s;
+- (void) parseFor:(NSString*)predicate On:(id)o;
+- (id) match:(NSString*)predicate On:(id)o;
+
+/*
+ * NOTE :
+ * Call the createManipulator method of a derived class, see
+ * for example OpalFuzzyPaintRuleSystem.h
+ */ 
+- (id)createManipulator;
+- (id)createArgumentManipulator;
+
+@end
+
+#endif
diff -Naur gnustep-back/Source/opal/GNUmakefile gnustep-back-work/Source/opal/GNUmakefile
--- gnustep-back/Source/opal/GNUmakefile	2014-12-15 12:26:51.000000000 +0100
+++ gnustep-back-work/Source/opal/GNUmakefile	2014-12-15 07:09:01.000000000 +0100
@@ -40,7 +40,9 @@
   ../fontconfig/FCFaceInfo.m \
   ../fontconfig/FCFontEnumerator.m \
   ../fontconfig/FCFontInfo.m \
-
+  fuzzy/OpalFuzzy.m \
+  fuzzy/OpalFuzzyDB.m \
+ 
 ifeq ($(BUILD_SERVER),x11DISABLED)
   ifeq ($(WITH_GLITZ),yes)
     opal_OBJC_FILES += XGCairoGlitzSurface.m
diff -Naur gnustep-back/Source/opal/GNUmakefile.orig gnustep-back-work/Source/opal/GNUmakefile.orig
--- gnustep-back/Source/opal/GNUmakefile.orig	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/GNUmakefile.orig	2014-12-12 16:47:19.000000000 +0100
@@ -0,0 +1,70 @@
+#  Copyright (C) 2013 Free Software Foundation, Inc.
+#
+#  Author: Ivan Vucica <ivan@vucica.net>
+#
+#  This file is part of the GNUstep Backend.
+#
+#  This library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public
+#  License as published by the Free Software Foundation; either
+#  version 2 of the License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library; see the file COPYING.LIB.
+#  If not, see <http://www.gnu.org/licenses/> or write to the 
+#  Free Software Foundation, 51 Franklin Street, Fifth Floor, 
+#  Boston, MA 02110-1301, USA.
+
+PACKAGE_NAME = gnustep-back
+GNUSTEP_LOCAL_ADDITIONAL_MAKEFILES=../../back.make
+include $(GNUSTEP_MAKEFILES)/common.make
+
+include ../../config.make
+
+SUBPROJECT_NAME=opal
+
+# The Objective-C source files to be compiled
+opal_OBJC_FILES = OpalSurface.m \
+  OpalFontInfo.m \
+  OpalGState.m \
+  OpalContext.m \
+  OpalFontEnumerator.m \
+  OpalFaceInfo.m \
+  OpalPSSurface.m \
+  OpalPDFSurface.m \
+  ../fontconfig/FCFaceInfo.m \
+  ../fontconfig/FCFontEnumerator.m \
+  ../fontconfig/FCFontInfo.m \
+  fuzzy/OpalFuzzy.m \
+  fuzzy/OpalFuzzyDB.m \
+ 
+ifeq ($(BUILD_SERVER),x11DISABLED)
+  ifeq ($(WITH_GLITZ),yes)
+    opal_OBJC_FILES += XGCairoGlitzSurface.m
+  else
+    opal_OBJC_FILES += XGCairoSurface.m XGCairoXImageSurface.m XGCairoModernSurface.m
+  endif
+else
+  ifeq ($(BUILD_GRAPHICS),opalDISABLED)
+    ifeq ($(WITH_GLITZ),yes)
+      opal_OBJC_FILES += Win32CairoGlitzSurface.m
+    else
+      opal_OBJC_FILES += Win32CairoSurface.m Win32CairoGState.m
+      # Win32CairoXImageSurface.m 
+    endif
+  endif
+endif
+
+opal_OBJC_FILES += 
+
+-include GNUmakefile.preamble
+
+include $(GNUSTEP_MAKEFILES)/subproject.make
+
+-include GNUmakefile.postamble
+
diff -Naur gnustep-back/Source/opal/GNUmakefile.rej gnustep-back-work/Source/opal/GNUmakefile.rej
--- gnustep-back/Source/opal/GNUmakefile.rej	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/GNUmakefile.rej	2014-12-15 07:09:01.000000000 +0100
@@ -0,0 +1,19 @@
+***************
+*** 40,46 ****
+    ../fontconfig/FCFaceInfo.m \
+    ../fontconfig/FCFontEnumerator.m \
+    ../fontconfig/FCFontInfo.m \
+- 
+  
+  
+  
+--- 40,48 ----
+    ../fontconfig/FCFaceInfo.m \
+    ../fontconfig/FCFontEnumerator.m \
+    ../fontconfig/FCFontInfo.m \
++   fuzzy/OpalFuzzy.m \
++   fuzzy/OpalFuzzyDB.m \
++  
+  
+  
+  
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzy.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzy.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzy.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzy.m	2014-12-15 09:20:51.000000000 +0100
@@ -0,0 +1,159 @@
+ /*
+    OpalFuzzy.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org>
+    Date: Dec 2014 
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #import "opal/fuzzy/OpalFuzzy.h"
+ 
+ @implementation OpalFuzzyFunction
+ 
+ - (id)new:(SEL)pif
+ {
+ 	[super alloc];
+ 	pf = pif;
+ 	x = .0;	
+ 	return self;
+ } 
+ 
+ -(void)perform:(id)c
+ {
+ 	[c performSelector:pf]
+ }
+ 
+ -(void)perform
+ {
+ 	[self performSelector:pf]
+ }
+ 
+ - (long int)x
+ {
+ 	return x;
+ }
+ 
+ - (void)x:(long int)xx
+ {
+ 	x = xx;
+ }
+ @end
+ 
+ @implementation OpalFuzzyFunctionLittle
+ 
+ -(void) littlef
+ {
+ 	return sqrt(x) / 1.3;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(littlef)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionSlightly
+ 
+ -(void) slightlyf
+ {
+ 	return sqrt(x) / 1.7;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(slightlyf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionVery
+ 
+ -(void) veryf
+ {
+ 	return x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(veryf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionExtremely
+ 
+ -(void) extremelyf
+ {
+ 	return x * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(extremelyf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionVeryVery
+ 
+ -(void) veryveryf
+ {
+ 	return x * x * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(veryveryf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionMoreorLess
+ 
+ -(void) moreorlessf
+ {
+ 	return sqrt(x);
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(moreorlessf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionIndeed
+ 
+ -(void) indeedf
+ {
+ 	return 2 * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(indeedf)];
+ }
+ 
+ @end
+ 
+ 
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyDB.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyDB.m	2014-12-15 09:30:07.000000000 +0100
@@ -0,0 +1,49 @@
+ /*
+    OpalFuzzyDB.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ @implementation OpalFuzzyDB {
+ 
+ -(id)new {
+ 	[[super alloc] init];
+ }
+ 
+ - (void)addToDict:(NSString*)key :(SEL)pif
+ {
+ 	[self setObject:pif forKey:key];
+ }
+ 
+ 
+ - (SEL)getWithKey:(NSString*)key
+ {
+ 	return [self objectForKey:key];	
+ }
+ 
+- (void)reinitCapacity
+{
+}
+
+ @end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyManipulator.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyManipulator.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyManipulator.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyManipulator.m	2014-12-15 12:16:49.000000000 +0100
@@ -0,0 +1,81 @@
+/*
+   OpalFuzzyManipulator.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyManipulator.h"
+
+
+@implementation OpalFuzzyScreenManipulator
+
+- (id) new:(id)screen
+{
+	[super new];
+	_o = screen; 
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyArgumentManipulator
+
+- (id) new:(id)o
+{
+	[[super alloc] init];
+	_o = o;
+	return self;
+}
+
+- (id)o
+{
+	return _o;
+}
+@end
+
+@implementation OpalFuzzyPaintManipulator
+
+- (id) new:(id)painter
+{
+	[super new];
+	_o = painter
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyManipulator
+
+- (id) new:(id)o
+{
+	[[super alloc] init];
+	_o = o;
+	return self;
+}
+
+- (id)o
+{
+	return _o;
+}
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyMatrix.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyMatrix.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyMatrix.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyMatrix.m	2014-12-15 08:44:38.000000000 +0100
@@ -0,0 +1,96 @@
+/*
+   OpalFuzzyMatrix.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyMatrix.h"
+
+@implementation OpalFuzzyMatrix
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_w = 1;
+	_h = 1;
+
+	MAKEMatrix(1,1);
+	[_matrix alloc];
+
+	return self;
+}
+
+- (id) setWidth:(long int)w height:(long int)h
+{
+       _w = w;
+       _h = h;
+
+	if (w > 0 && h > 0) {
+		MAKEMatrix(_w,_h);
+		[_matrix alloc];
+	} else {
+		NSLog(@"Could not typedef matrix structure as width or height is <= 0, _matrix width==1  and height==1 from new, returning nil");
+		return nil;
+	}
+	
+	return self;	
+}
+
+- (BOOL)hasDimensionsBeenSet:(BOOL)b
+{
+       if (!b) {
+               if (_w <= 0 or _h <= 0) {
+                       [self _calibrateDimensions];
+                       return NO;
+               } else {
+                       return YES;
+               }
+       } else {
+               if (_w <= 0 or _h <= 0) {
+                       [self _calibrateDimensions];
+                       return NO;
+               } else {
+                       return YES;
+               }
+
+       }
+}
+
+// private methods
+
+- (void) calibrateDimensions
+{
+       /* FIXME */
+       /***
+	_w = 1;
+       _h = 1;
+	MAKEMatrix(_w,_h);
+	[_matrix alloc];
+
+	***/
+
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyNetwork.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNetwork.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyNetwork.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNetwork.m	2014-12-15 07:09:01.000000000 +0100
@@ -0,0 +1,43 @@
+/*
+   OpalFuzzyNetwork.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyNetwork.h"
+
+
+@implementation OpalFuzzyNetwork
+
+- (id) new
+{
+	[[super alloc] init];
+	return self;
+}
+
+- 
+
+@end
+
+
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyNode.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNode.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyNode.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNode.m	2014-12-15 10:52:07.000000000 +0100
@@ -0,0 +1,53 @@
+/*
+   OpalFuzzyNode.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyNode.h"
+
+
+@implementation OpalFuzzyPaintNode
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_rules = [OpalFuzzyPaintRuleSystem new];
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyNode
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_rules = [OpalFuzzyRuleSystem new];
+	return self;
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m	2014-12-15 12:22:15.000000000 +0100
@@ -0,0 +1,70 @@
+/*
+   OpalFuzzyPaintRuleSystem.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyPaintRuleSystem.h"
+
+
+@implementation OpalFuzzyPaintRuleSystem
+
+- (id) new:(id)painter
+{
+	[super new];
+	_painter = painter;
+	return self;
+}
+
+- (void) setPaintSystem:(id)paintsystem
+{
+	_painter = paintsystem;
+}
+
+-(void)painter
+{
+	return _painter;
+}
+
+- (void) paintOn
+{
+	/* NOTE : painting procedures without dispatch, use a pattern */
+}
+
+- (void) paintOn:(id)painter
+{
+	
+}
+
+- (id) createManipulator
+{
+	return [OpalFuzzyPaintManipulator new:self];
+}
+
+- (id) createArgumentManipulator
+{
+	return [OpalFuzzyScreenManipulator new:_painter];
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyRuleSystem.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyRuleSystem.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyRuleSystem.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyRuleSystem.m	2014-12-15 12:14:35.000000000 +0100
@@ -0,0 +1,79 @@
+/*
+   OpalFuzzyRuleSystem.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyDB.h"
+#import "opal/fuzzy/OpalFuzzyRuleSystem.h"
+
+
+@implementation OpalFuzzyRuleSystem
+
+- (id) new
+{
+	[[super alloc] init];
+	_rules = [[NSMutableArray alloc] initWithCapacity:0];
+	return self;
+}
+
+- (void) addRule:(NSString*)r with:(SEL)s
+{
+	[_rules reinitCapacity];	
+	[_rules addToDict:r :s];
+}
+
+-(void) parseFor:(NSString*)predicate On:(id)o
+{
+	key = [self match:predicate On:o];
+	SEL s = [_rules objectForKey:key];	
+}
+
+- (id) match:(NSString*)predicate On:(id)o
+{
+	/* FIXME better rule matching eg more rules than first found */
+    NSEnumerator *enumerator = [_rules keyEnumerator];
+    id key;
+    while ((key = [enumerator nextObject])) {
+	if ([predicate rangeOfString:key]->length == 0)
+			continue;
+		else {
+			[self performSelector:(SEL)[_rules objectForKey:key] withObject:o];
+			return key;
+		}	
+	}	
+
+}	
+
+- (id) createManipulator
+{
+	return [OpalFuzzyManipulator new:self];
+}
+
+- (id) createArgumentManipulator
+{
+	return [OpalFuzzyArgumentManipulator new:self];
+}
+
