diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzy.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzy.h	2014-12-15 14:07:11.000000000 +0100
@@ -0,0 +1,91 @@
+ /*
+    OpalFuzzy.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 021101301, USA.
+ */
+ 
+ #ifndef OpalFuzzy_h_defined
+ #define OpalFuzzy_h_defined
+
+
+//#define IDLEVALUE(x)\
+//	return 0
+ 
+ @interface OpalFuzzyFunction : NSObject {
+ 	SEL pf;
+ 	CGFloat x;
+ }
+-  (id) new:(SEL) pif; 
+- (CGFloat)perform:(CGFloat)c;
+- (CGFloat)perform;
+- (CGFloat)x;
+- (void)x:(CGFloat)xx;
+- (SEL)sel;
+ @end
+ 
+ @interface OpalFuzzyFunctionLittle : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionSlightly : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionVery : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionExtremely : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionVeryVery : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionMoreorLess : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionSomewhat : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionIndeed : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ @interface OpalFuzzyFunctionIdle : OpalFuzzyFunction {
+ }
+  (id) new; 
+ @end
+ 
+ @interface OpalFuzzy : NSObject {
+ 	
+ }
+ 
+ 
+ @end
+ 
+ #endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyDB.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyDB.h	2014-12-15 14:09:37.000000000 +0100
@@ -0,0 +1,41 @@
+ /*
+    OpalFuzzyDB.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #ifndef OpalFuzzyDB_h_defined
+ #define OpalFuzzyDB_h_defined
+ 
+ @interface OpalFuzzyDB : NSMutableDictionary {
+ }
+ - (id)new;
+ - (void)addSelectorToDict:(NSString*)key :(SEL)pif;
+ - (void)addObjectToDict:(NSString*)key :(id)o;
+ - (SEL)getWithKey:(NSString*)key;
+ - (void)reinitCapacity;
+ 
+ @end
+ 
+ #endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyManipulator.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyManipulator.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyManipulator.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyManipulator.h	2014-12-15 16:56:44.000000000 +0100
@@ -0,0 +1,68 @@
+/*
+   OpalFuzzyManipulator.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyManipulator_h_defined
+#define OpalFuzzyManipulator_h_defined
+
+@interface OpalFuzzyParserManipulator : OpalFuzzyManipulator
+{
+}
+
+- (id) new:(id)parser;
+@end
+
+@interface OpalFuzzyScreenManipulator : OpalFuzzyArgumentManipulator
+{
+}
+
+- (id) new:(id)screen;
+@end
+
+@interface OpalArgumentManipulator : NSObject 
+{
+	id _o;
+}
+- (id) new:(id)o;
+- (id) o;
+@end
+
+@interface OpalFuzzyPaintManipulator : OpalFuzzyManipulator
+{
+}
+
+- (id) new:(id)painter;
+@end
+
+@interface OpalFuzzyManipulator : NSObject
+{
+	id _o;
+}
+- (id) new:(id)o;
+- (id) o;
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyMatrix.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyMatrix.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyMatrix.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyMatrix.h	2014-12-15 16:15:55.000000000 +0100
@@ -0,0 +1,56 @@
+/*
+   OpalFuzzyMatrix.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyMatrix_h_defined
+#define OpalFuzzyMatrix_h_defined
+
+#define MAKEMatrixElement(i,j) \
+	OpalFuzzyNode * c ## #i#j;
+
+/* FIXME extra handle or not */
+#define MAKEMatrix(w, h) \
+	typedef struct {
+		for (i = 0; i < h; i++) \
+			for (j = 0; j < w; j++) \
+				makeMatrixElement(i,j) \
+} _OpalMatrix;		
+
+@class _OpalMatrix;
+
+@interface OpalFuzzyMatrix : NSObject {
+	long int _w,_h;
+	_OpalMatrix *_matrix;	
+}
+
+- (id) new;
+- (id)setWidth:(long int)w height:(long int)h;
+- (BOOL)hasDimensionsBeenSet:(BOOL)b;
+- (void) calibrateDimensions;
+ 
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyNetwork.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNetwork.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyNetwork.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNetwork.h	2014-12-15 07:08:58.000000000 +0100
@@ -0,0 +1,37 @@
+/*
+   OpalFuzzyNetwork.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyNetwork_h_defined
+#define OpalFuzzyNetwork_h_defined
+
+@interface OpalFuzzyNetwork : NSObject {
+}
+- (id) new;
+ 
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyNode.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNode.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyNode.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyNode.h	2014-12-15 10:51:27.000000000 +0100
@@ -0,0 +1,44 @@
+/*
+   OpalFuzzyNode.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyNode_h_defined
+#define OpalFuzzyNode_h_defined
+
+@class OpalFuzzyRuleSystem;
+
+@interface OpalFuzzyPaintNode : OpalFuzzyNode
+{}
+
+@end
+
+@interface OpalFuzzyNode : NSObject
+{
+	OpalFuzzyRuleSystem *_rules;
+}
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintParser.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintParser.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintParser.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintParser.h	2014-12-15 16:39:46.000000000 +0100
@@ -0,0 +1,40 @@
+ /*
+    OpalFuzzyPaintParser.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #ifndef OpalFuzzyPaintParser_h_defined
+ #define OpalFuzzyPaintParser_h_defined
+
+@class OpalFuzzyParser;
+ 
+@interface OpalFuzzyPaintParser : OpalFuzzyParser {
+}
+- (id)new;
+- (id) compare:(NSString*)predicate with:(NSString*)rule;
+ 
+@end
+ 
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRuleSystem.h	2014-12-15 17:04:10.000000000 +0100
@@ -0,0 +1,71 @@
+/*
+   OpalFuzzyPaintRuleSystem.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyPaintRuleSystem_h_defined
+#define OpalFuzzyPaintRuleSystem_h_defined
+
+@class OpalFuzzyRuleSystem;
+@class OpalFuzzyParser;
+@class OpalFuzzyPredicate;
+
+@interface OpalFuzzyPaintRuleSystem : OpalFuzzyRuleSystem 
+{
+@private
+	id _painter;
+	id _parser;
+}
+- (id) new:(id)painter withParser:(id)parser;
+- (void)setPaintSystem:(id)paintsystem;
+- (void)painter;
+- (void)setParser:(id)parser;
+- (void)parser;
+- (void)paintOn;
+
+- (void)paintOn:(id)painter;
+- (void)paintIdle1SecondOn:(id)painter;
+
+- (id) match:(OpalFuzzyPredicate*)predicate On:(id)o;
+/*
+ * NOTE :
+ * Call the createManipulator method and you can paint on this
+ * painter system with rules from the super class.
+
+ * The PaintManipulator then hashes the method paintOn: in the dictionary
+ * and paints on the _painter of the dispatched class due to a call
+ * of e.g. parseFor:"paint extremely". 
+ *
+ * The super class parseFor:On: method can also be used as a rule (rules have
+ * a NSString key and a SEL in the form of performSelector:withObject:, so
+ * only a single argument.
+ */ 
+- (id) createManipulator;
+- (id) createArgumentManipulator;
+- (id) createParserManipulator;
+
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRules.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRules.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyPaintRules.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPaintRules.h	2014-12-15 13:07:15.000000000 +0100
@@ -0,0 +1,33 @@
+ /*
+    OpalFuzzyPaintRules.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 021101301, USA.
+ */
+ 
+#ifndef OpalFuzzyPaintRules_h_defined
+#define OpalFuzzyPaintRules_h_defined
+
+void makePaintRules((OpalFuzzyPaintRuleSystem*)sys);
+
+#endif 
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyParser.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyParser.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyParser.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyParser.h	2014-12-15 16:41:37.000000000 +0100
@@ -0,0 +1,38 @@
+ /*
+    OpalFuzzyParser.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+#ifndef OpalFuzzyParser_h_defined
+#define OpalFuzzyParser_h_defined
+ 
+@interface OpalFuzzyParser : NSObject {
+}
+- (id)new;
+- (id) compare:(NSString*)predicate with:(NSString*)rule;
+ 
+@end
+ 
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyPredicate.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPredicate.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyPredicate.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyPredicate.h	2014-12-15 15:22:33.000000000 +0100
@@ -0,0 +1,37 @@
+ /*
+    OpalFuzzyPredicate.h
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #ifndef OpalFuzzyPredicate_h_defined
+ #define OpalFuzzyPredicate_h_defined
+ 
+ @interface OpalFuzzyPredicate : NSString {
+ }
+- (id)compareWithParser:(OpalFuzzyParser*)parser and:(OpalFuzzyPredicate*)rule;
+ 
+ @end
+ 
+ #endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Headers/opal/fuzzy/OpalFuzzyRuleSystem.h	2014-12-15 16:44:45.000000000 +0100
@@ -0,0 +1,56 @@
+/*
+   OpalFuzzyRuleSystem.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyRuleSystem_h_defined
+#define OpalFuzzyRuleSystem_h_defined
+
+@class OpalFuzzyDB;
+@class OpalFuzzyPredicate;
+
+@interface OpalFuzzyRuleSystem : NSObject
+{
+	CGFloat _threshold;
+	OpalFuzzyDB *_rules;
+	OpalFuzzyDB *_functions;
+}
+- (id) new;
+- (void) addRule:(NSString*)r with:(SEL)s;
+- (void) addFunction:(NSString*)r with:(OpalFuzzyFunction*)s;
+- (void) parseFor:(OpalFuzzyPredicate*)predicate On:(id)o;
+- (id) match:(OpalFuzzyPredicate*)predicate On:(id)o;
+
+/*
+ * NOTE :
+ * Call the createManipulator method of a derived class, see
+ * for example OpalFuzzyPaintRuleSystem.h
+ */ 
+- (id)createManipulator;
+- (id)createArgumentManipulator;
+
+@end
+
 
diff -Naur gnustep-back/Source/opal/GNUmakefile gnustep-back-work/Source/opal/GNUmakefile
--- gnustep-back/Source/opal/GNUmakefile	2014-12-15 12:26:51.000000000 +0100
+++ gnustep-back-work/Source/opal/GNUmakefile	2014-12-15 16:53:02.000000000 +0100
@@ -40,7 +40,19 @@
   ../fontconfig/FCFaceInfo.m \
   ../fontconfig/FCFontEnumerator.m \
   ../fontconfig/FCFontInfo.m \
-
+  fuzzy/OpalFuzzy.m \
+  fuzzy/OpalFuzzyDB.m \
+  fuzzy/OpalFuzzyManipulator.m \
+  fuzzy/OpalFuzzyMatrix.m \
+  fuzzy/OpalFuzzyNetwork.m \
+  fuzzy/OpalFuzzyNode.m \
+  fuzzy/OpalFuzzyPaintParser.m \
+  fuzzy/OpalFuzzyPaintRuleSystem.m \
+  fuzzy/OpalFuzzyPaintRules.m \
+  fuzzy/OpalFuzzyParser.m \
+  fuzzy/OpalFuzzyPredicate.m \
+  fuzzy/OpalFuzzyRuleSystem.m
+ 
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzy.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzy.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzy.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzy.m	2014-12-15 14:06:55.000000000 +0100
@@ -0,0 +1,179 @@
+ /*
+    OpalFuzzy.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org>
+    Date: Dec 2014 
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ #import "opal/fuzzy/OpalFuzzy.h"
+ 
+ @implementation OpalFuzzyFunction
+ 
+ - (id)new:(SEL)pif
+ {
+ 	[super alloc];
+ 	pf = pif;
+ 	x = .0;	
+ 	return self;
+ } 
+ 
+ -(CGFloat)perform:(CGFloat)c
+ {
+	x = c;
+ 	return [self performSelector:pf];
+ }
+ 
+ -(CGFloat)perform
+ {
+ 	return [self performSelector:pf];
+ }
+ 
+ - (CGFloat)x
+ {
+ 	return x;
+ }
+ 
+ - (void)x:(CGFloat)xx
+ {
+ 	x = xx;
+ }
+ 
+ - (SEL)sel
+ {
+	return pf;
+ }
+ @end
+ 
+ @implementation OpalFuzzyFunctionLittle
+ 
+ -(CGFloat) littlef
+ {
+ 	return sqrt(x) / 1.3;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(littlef)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionSlightly
+ 
+ -(CGFloat) slightlyf
+ {
+ 	return sqrt(x) / 1.7;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(slightlyf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionVery
+ 
+ -(CGFloat) veryf
+ {
+ 	return x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(veryf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionExtremely
+ 
+ -(CGFloat) extremelyf
+ {
+ 	return x * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(extremelyf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionVeryVery
+ 
+ -(CGFloat) veryveryf
+ {
+ 	return x * x * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(veryveryf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionMoreorLess
+ 
+ -(CGFloat) moreorlessf
+ {
+ 	return sqrt(x);
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(moreorlessf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionIndeed
+ 
+ -(CGFloat) indeedf
+ {
+ 	return 2 * x * x;
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(indeedf)];
+ }
+ 
+ @end
+ 
+ @implementation OpalFuzzyFunctionIdle
+ 
+ -(CGFloat) idlef
+ {
+ 	return .0; //IDLEVALUE(x);
+ }
+ 
+ -(id) new 
+ {
+ 	[self new:@selector(idlef)];
+ }
+ 
+ @end
+ 
+ 
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyDB.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyDB.m	2014-12-15 14:45:43.000000000 +0100
@@ -0,0 +1,55 @@
+ /*
+    OpalFuzzyDB.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ @implementation OpalFuzzyDB {
+ 
+ -(id)new {
+ 	[[super alloc] init];
+	return self; 
+}
+ 
+ - (void)addSelectorToDict:(NSString*)key :(SEL)pif
+ {
+ 	[self setObject:pif forKey:key];
+ }
+ 
+ - (void)addObjectToDict:(NSString*)key :(id)o
+ {
+ 	[self setObject:o forKey:key];
+ }
+ 
+ 
+ - (SEL)getWithKey:(NSString*)key
+ {
+ 	return [self objectForKey:key];	
+ }
+ 
+- (void)reinitCapacity
+{
+}
+
+ @end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyManipulator.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyManipulator.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyManipulator.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyManipulator.m	2014-12-15 16:57:21.000000000 +0100
@@ -0,0 +1,92 @@
+/*
+   OpalFuzzyManipulator.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyManipulator.h"
+
+
+@implementation OpalFuzzyParserManipulator
+
+- (id) new:(id)parser
+{
+	[super new];
+	_o = parser; 
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyScreenManipulator
+
+- (id) new:(id)screen
+{
+	[super new];
+	_o = screen; 
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyArgumentManipulator
+
+- (id) new:(id)o
+{
+	[[super alloc] init];
+	_o = o;
+	return self;
+}
+
+- (id)o
+{
+	return _o;
+}
+@end
+
+@implementation OpalFuzzyPaintManipulator
+
+- (id) new:(id)painter
+{
+	[super new];
+	_o = painter
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyManipulator
+
+- (id) new:(id)o
+{
+	[[super alloc] init];
+	_o = o;
+	return self;
+}
+
+- (id)o
+{
+	return _o;
+}
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyMatrix.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyMatrix.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyMatrix.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyMatrix.m	2014-12-15 08:44:38.000000000 +0100
@@ -0,0 +1,96 @@
+/*
+   OpalFuzzyMatrix.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyMatrix.h"
+
+@implementation OpalFuzzyMatrix
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_w = 1;
+	_h = 1;
+
+	MAKEMatrix(1,1);
+	[_matrix alloc];
+
+	return self;
+}
+
+- (id) setWidth:(long int)w height:(long int)h
+{
+       _w = w;
+       _h = h;
+
+	if (w > 0 && h > 0) {
+		MAKEMatrix(_w,_h);
+		[_matrix alloc];
+	} else {
+		NSLog(@"Could not typedef matrix structure as width or height is <= 0, _matrix width==1  and height==1 from new, returning nil");
+		return nil;
+	}
+	
+	return self;	
+}
+
+- (BOOL)hasDimensionsBeenSet:(BOOL)b
+{
+       if (!b) {
+               if (_w <= 0 or _h <= 0) {
+                       [self _calibrateDimensions];
+                       return NO;
+               } else {
+                       return YES;
+               }
+       } else {
+               if (_w <= 0 or _h <= 0) {
+                       [self _calibrateDimensions];
+                       return NO;
+               } else {
+                       return YES;
+               }
+
+       }
+}
+
+// private methods
+
+- (void) calibrateDimensions
+{
+       /* FIXME */
+       /***
+	_w = 1;
+       _h = 1;
+	MAKEMatrix(_w,_h);
+	[_matrix alloc];
+
+	***/
+
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyNetwork.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNetwork.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyNetwork.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNetwork.m	2014-12-15 07:09:01.000000000 +0100
@@ -0,0 +1,43 @@
+/*
+   OpalFuzzyNetwork.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyNetwork.h"
+
+
+@implementation OpalFuzzyNetwork
+
+- (id) new
+{
+	[[super alloc] init];
+	return self;
+}
+
+- 
+
+@end
+
+
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyNode.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNode.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyNode.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyNode.m	2014-12-15 12:56:04.000000000 +0100
@@ -0,0 +1,53 @@
+/*
+   OpalFuzzyNode.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyNode.h"
+
+
+@implementation OpalFuzzyPaintNode
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_rules = [OpalFuzzyPaintRuleSystem new];
+	return self;
+}
+
+@end
+
+@implementation OpalFuzzyNode
+
+- (id) new
+{
+	[[super alloc] init];
+
+	_rules = [OpalFuzzyRuleSystem new];
+	return self;
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintParser.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintParser.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintParser.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintParser.m	2014-12-15 16:40:56.000000000 +0100
@@ -0,0 +1,57 @@
+ /*
+    OpalFuzzyPaintParser.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+#import "opal/fuzzy/OpalFuzzyPaintParser.h"
+ 
+ @implementation OpalFuzzyPaintParser {
+ 
+ -(id)new {
+ 	[[super alloc] init];
+	return self; 
+}
+
+/*
+ * NOTE :
+ * compare function below parses the non-fuzzy part of a predicate 
+ *
+ */
+
+- (id) compare:(NSString*)predicate with:(NSString*)rule
+{
+	/* perfect match */
+	if ([predicate rangeOfString:rule]->length == [rule length]) {
+		return rule;
+	}
+	/* parse non-fuzzy predicate stuff */
+	if ([predicate rangeOfString:"paint"]->length == 5)
+		return rule;
+	else if ([predicate rangeOfString:"second"]->length == 6)
+		return rule;
+	else
+		return [NSString alloc];
+}
+
+@end 
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRuleSystem.m	2014-12-15 17:05:09.000000000 +0100
@@ -0,0 +1,114 @@
+/*
+   OpalFuzzyPaintRuleSystem.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyManipulator.h"
+#import "opal/fuzzy/OpalFuzzyPaintRuleSystem.h"
+#import "opal/fuzzy/OpalFuzzyPaintParser.h"
+
+@implementation OpalFuzzyPaintRuleSystem
+
+- (id) new:(id)painter andParser:(id)parser
+{
+	[super new];
+	_painter = painter;
+	_parser = parser;
+	return self;
+}
+
+- (void) setPaintSystem:(id)paintsystem
+{
+	_painter = paintsystem;
+}
+
+-(void)painter
+{
+	return _painter;
+}
+
+- (void) setParser:(id)parser
+{
+	_parser = parser;
+}
+
+-(void)parser
+{
+	return _parser;
+}
+
+- (void) paintOn
+{
+	/* NOTE : painting procedures without dispatch, use a pattern */
+}
+
+- (void) paintOn:(id)painter
+{
+	
+}
+
+- (void) paintIdle1SecondOn:(id)painter
+{
+	//thread or sleep(1);	
+}
+
+- (id) match:(OpalFuzzyPredicate*)predicate On:(id)o
+{
+    NSEnumerator *enumerator = [_rules keyEnumerator];
+    id key;
+    while ((key = [enumerator nextObject])) {
+	/* greedy full match rule search */
+	if ([predicate rangeOfString:key]->length == 0)
+			continue;
+	else if ([[predicate compareWithParser:[self parser] and: key] length] > 0) {
+		/* if key found, perform fuzzy action and set arg with x:
+		 before */
+		if ([[_functions objectForKey:key] perform] < _threshold) { 
+			[self performSelector:(SEL)[_rules objectForKey:key] withObject:o];
+			return key;
+		}
+	} else {
+		return nil;
+	}	
+     }	
+
+}	
+
+- (id) createManipulator
+{
+	return [OpalFuzzyPaintManipulator new:self];
+}
+
+- (id) createArgumentManipulator
+{
+	return [OpalFuzzyScreenManipulator new:_painter];
+}
+
+- (id) createParserManipulator
+{
+	return [OpalFuzzyParserManipulator new:_parser];
+}
+
+@end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRules.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRules.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyPaintRules.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPaintRules.m	2014-12-15 16:17:47.000000000 +0100
@@ -0,0 +1,42 @@
+ /*
+    OpalFuzzy.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org>
+    Date: Dec 2014 
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+#import "opal/fuzzy/OpalFuzzyPaintRules.h"
+
+void makePaintRules(OpalFuzzyPaintRuleSystem* sys)
+{
+	[sys addRule:"paint extremely" : paintOn:];
+	OpalFuzzyFunction *f = [OpalFuzzyFunctionExtremely new];
+	[f x:.0];
+	[sys addFunction:"paint extremely" : f];
+
+	[sys addRule:"paint not for 1 second" : paintIdle1SecondOn:];
+	f = [OpalFuzzyFunctionIdle new];
+	[f x:.0];
+	[sys addFunction:"paint not for 1 second" : f];
+}
+
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyParser.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyParser.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyParser.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyParser.m	2014-12-15 16:41:22.000000000 +0100
@@ -0,0 +1,40 @@
+ /*
+    OpalFuzzyParser.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ @implementation OpalFuzzyParser {
+ 
+ -(id)new {
+ 	[[super alloc] init];
+	return self; 
+}
+
+- (id) compare:(NSString*)predicate with:(NSString*)rule
+{
+	return nil; 
+}
+
+@end 
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyPredicate.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPredicate.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyPredicate.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyPredicate.m	2014-12-15 15:22:17.000000000 +0100
@@ -0,0 +1,34 @@
+ /*
+    OpalFuzzyPredicate.m
+ 
+    Copyright (C) 2014 Free Software Foundation, Inc.
+ 
+    Author: Johan Ceuppens <johan@yellowcouch.org> 
+    Date: Dec 2014
+ 
+    This file is part of GNUstep.
+ 
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+ 
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+    Lesser General Public License for more details.
+ 
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; see the file COPYING.LIB.
+    If not, see <http://www.gnu.org/licenses/> or write to the
+    Free Software Foundation, 51 Franklin Street, Fifth Floor,
+    Boston, MA 02110-1301, USA.
+ */
+ 
+ @implementation OpalFuzzyPredicate {
+
+- (id)compareWithParser:(OpalFuzzyParser*)parser and:(OpalFuzzyPredicate*)rule
+{
+	return [_parser compare:self with:rule];
+} 
+ @end
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyRuleSystem.m gnustep-back-work/Source/opal/fuzzy/OpalFuzzyRuleSystem.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyRuleSystem.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-work/Source/opal/fuzzy/OpalFuzzyRuleSystem.m	2014-12-15 16:44:58.000000000 +0100
@@ -0,0 +1,91 @@
+/*
+   OpalFuzzyRuleSystem.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzyPredicate.h"
+#import "opal/fuzzy/OpalFuzzyDB.h"
+#import "opal/fuzzy/OpalFuzzyRuleSystem.h"
+
+
+@implementation OpalFuzzyRuleSystem
+
+- (id) new
+{
+	[[super alloc] init];
+	_threshold = 0.001;	
+	_rules = [OpalFuzzyDB new]; 
+	_functions = [OpalFuzzyDB new]; 
+	return self;
+}
+
+- (void) addRule:(NSString*)r with:(SEL)s
+{
+	[_rules reinitCapacity];	
+	[_rules addSelectorToDict:r :s];
+}
+
+- (void) addFunction:(NSString*)r with:(OpalFuzzyFunction*)f
+{
+	[_functions reinitCapacity];	
+	[_functions addObjectToDict:r :f];
+}
+
+-(void) parseFor:(OpalFuzzyPredicate*)predicate On:(id)o
+{
+	NSString *key = [self match:predicate On:o];
+}
+
+- (id) match:(OpalFuzzyPredicate*)predicate On:(id)o
+{
+    NSEnumerator *enumerator = [_rules keyEnumerator];
+    id key;
+    while ((key = [enumerator nextObject])) {
+	/* greedy full match rule search */
+	if ([predicate rangeOfString:key]->length == 0)
+			continue;
+	else if ([predicate rangeOfString: key]->length > 0) {
+		/* if key found, perform fuzzy action on f(x) with x = 0 */
+		if ([[_functions objectForKey:key] perform] < _threshold) { 
+			[self performSelector:(SEL)[_rules objectForKey:key] withObject:o];
+			return key;
+		}
+	} else {
+		return nil;
+	}	
+     }	
+
+}	
+- (id) createManipulator
+{
+	return [OpalFuzzyManipulator new:self];
+}
+
+- (id) createArgumentManipulator
+{
+	return [OpalFuzzyArgumentManipulator new:self];
+}
+
+@end
