diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h gnustep-back-todiffdir/Headers/opal/fuzzy/OpalFuzzy.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzy.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-todiffdir/Headers/opal/fuzzy/OpalFuzzy.h	2014-12-12 16:41:31.000000000 +0100
@@ -0,0 +1,80 @@
+/*
+   OpalFuzzy.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzy_h_defined
+#define OpalFuzzy_h_defined
+
+@interface OpalFuzzyFunction : NSObject {
+	SEL pf;
+	double x;
+}
+- (id) new:(SEL) pif; 
+-(void)perform:(id)c;
+- (void) perform;
+@end
+
+@interface OpalFuzzyFunctionLittle : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionSlightly : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionVery : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionExtremely : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionVeryVery : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionMoreorLess : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionSomewhat : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+@interface OpalFuzzyFunctionIndeed : OpalFuzzyFunction {
+}
+- (id) new; 
+@end
+
+@interface OpalFuzzy : NSObject {
+	
+}
+
+
+@end
+
+#endif
diff -Naur gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h gnustep-back-todiffdir/Headers/opal/fuzzy/OpalFuzzyDB.h
--- gnustep-back/Headers/opal/fuzzy/OpalFuzzyDB.h	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-todiffdir/Headers/opal/fuzzy/OpalFuzzyDB.h	2014-12-12 16:41:31.000000000 +0100
@@ -0,0 +1,39 @@
+/*
+   OpalFuzzyDB.h
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#ifndef OpalFuzzyDB_h_defined
+#define OpalFuzzyDB_h_defined
+
+@interface OpalFuzzyDB : NSMutableDictionary {
+}
+- (id)new;
+- (void)addToDict:(NSString*)key :(SEL)pif;
+- (SEL)getWithKey:(NSString*)key;
+
+@end
+
+#endif
diff -Naur gnustep-back/Source/opal/GNUmakefile gnustep-back-todiffdir/Source/opal/GNUmakefile
--- gnustep-back/Source/opal/GNUmakefile	2014-12-12 16:40:46.000000000 +0100
+++ gnustep-back-todiffdir/Source/opal/GNUmakefile	2014-12-12 16:47:28.000000000 +0100
@@ -40,7 +40,9 @@
   ../fontconfig/FCFaceInfo.m \
   ../fontconfig/FCFontEnumerator.m \
   ../fontconfig/FCFontInfo.m \
-
+  fuzzy/OpalFuzzy.m \
+  fuzzy/OpalFuzzyDB.m \
+ 
 ifeq ($(BUILD_SERVER),x11DISABLED)
   ifeq ($(WITH_GLITZ),yes)
     opal_OBJC_FILES += XGCairoGlitzSurface.m
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzy.m gnustep-back-todiffdir/Source/opal/fuzzy/OpalFuzzy.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzy.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-todiffdir/Source/opal/fuzzy/OpalFuzzy.m	2014-12-12 16:47:28.000000000 +0100
@@ -0,0 +1,150 @@
+/*
+   OpalContext.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org>
+   Date: Dec 2014 
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+#import "opal/fuzzy/OpalFuzzy.h"
+
+@implementation OpalFuzzyFunction
+
+- (id)new:(SEL)pif
+{
+	[super alloc];
+	pf = pif;
+	x = .0;	
+	return self;
+} 
+
+-(void)perform:(id)c
+{
+	[c performSelector:pf]
+}
+
+-(void)perform
+{
+	[self performSelector:pf]
+}
+
+@end
+
+@implementation OpalFuzzyFunctionLittle
+
+-(void) littlef
+{
+	return sqrt(x) / 1.3;
+}
+
+-(id) new 
+{
+	[self new:@selector(littlef)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionSlightly
+
+-(void) slightlyf
+{
+	return sqrt(x) / 1.7;
+}
+
+-(id) new 
+{
+	[self new:@selector(slightlyf)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionVery
+
+-(void) veryf
+{
+	return x * x;
+}
+
+-(id) new 
+{
+	[self new:@selector(veryf)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionExtremely
+
+-(void) extremelyf
+{
+	return x * x * x;
+}
+
+-(id) new 
+{
+	[self new:@selector(extremelyf)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionVeryVery
+
+-(void) veryveryf
+{
+	return x * x * x * x;
+}
+
+-(id) new 
+{
+	[self new:@selector(veryveryf)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionMoreorLess
+
+-(void) moreorlessf
+{
+	return sqrt(x);
+}
+
+-(id) new 
+{
+	[self new:@selector(moreorlessf)];
+}
+
+@end
+
+@implementation OpalFuzzyFunctionIndeed
+
+-(void) indeedf
+{
+	return 2 * x * x;
+}
+
+-(id) new 
+{
+	[self new:@selector(indeedf)];
+}
+
+@end
+
+
diff -Naur gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m gnustep-back-todiffdir/Source/opal/fuzzy/OpalFuzzyDB.m
--- gnustep-back/Source/opal/fuzzy/OpalFuzzyDB.m	1970-01-01 01:00:00.000000000 +0100
+++ gnustep-back-todiffdir/Source/opal/fuzzy/OpalFuzzyDB.m	2014-12-12 16:47:28.000000000 +0100
@@ -0,0 +1,45 @@
+/*
+   OpalFuzzyDB.m
+
+   Copyright (C) 2014 Free Software Foundation, Inc.
+
+   Author: Johan Ceuppens <johan@yellowcouch.org> 
+   Date: Dec 2014
+
+   This file is part of GNUstep.
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; see the file COPYING.LIB.
+   If not, see <http://www.gnu.org/licenses/> or write to the
+   Free Software Foundation, 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+*/
+
+@implementation OpalFuzzyDB {
+
+-(id)new {
+	[[super alloc] init];
+}
+
+- (void)addToDict:(NSString*)key :(SEL)pif
+{
+	[self setObject:pif forKey:key];
+}
+
+
+- (SEL)getWithKey:(NSString*)key
+{
+	return [self objectForKey:key];	
+}
+
+@end
