Index: Source/NSPage.m
===================================================================
--- Source/NSPage.m	(revision 37522)
+++ Source/NSPage.m	(working copy)
@@ -63,6 +63,20 @@
 #define getpagesize()  B_PAGE_SIZE
 #endif
 
+#ifdef __ANDROID_API__
+/* Android is __linux__, but doesn't expose sysinfo() in libc */
+#include <sys/linux-syscalls.h>
+
+/* Linux developers promise not to break userland. Hopefully,
+   Android kernel developers have the same mindset. */
+#define __NR_sysinfo 116
+int
+sysinfo (struct sysinfo *info)
+{
+  return syscall(__NR_sysinfo, info);
+}
+#endif
+
 /* Cache the size of a memory page here, so we don't have to make the
    getpagesize() system call repeatedly. */
 static NSUInteger ns_page_size = 0;
